report_qweb_pdf_watermark/models/report.py

```diff
@@ -17,73 +17,66 @@
-        )
-        report = self._get_report_from_name(cr, uid, report_name)
-        watermark = None
+    @api.multi
+    def _scale_watermark(self, report, image):
+        size = parse_couple(
+            SIZES[report.paperformat_id.format.lower()])
+        size = int(size[0]), int(size[1])
+        return image.resize(size, resample=Image.LANCZOS)
+
+    @api.multi
+    def _read_watermark(self, report):
         if report.pdf_watermark:
             watermark = b64decode(report.pdf_watermark)
         else:
-            env = api.Environment(cr, uid, context)
             watermark = tools.safe_eval(
                 report.pdf_watermark_expression or 'None',
-                dict(env=env, docs=env[report.model].browse(ids)),
+                dict(
+                    env=self.env,
+                    docs=self.env[report.model].browse(self.ids),
+                )
             )
             if watermark:
                 watermark = b64decode(watermark)
-
         if not watermark:
-            return result
-
-        pdf = PdfFileWriter()
-        pdf_watermark = None
+            return
         try:
             pdf_watermark = PdfFileReader(StringIO(watermark))
         except PdfReadError:
-            # let's see if we can convert this with pillow
-            try:
-                image = Image.open(StringIO(watermark))
-                pdf_buffer = StringIO()
-                if image.mode != 'RGB':
-                    image = image.convert('RGB')
-                resolution = image.info.get(
-                    'dpi', report.paperformat_id.dpi or 90
-                )
-                if isinstance(resolution, tuple):
-                    resolution = resolution[0]
-                image.save(pdf_buffer, 'pdf', resolution=resolution)
-                pdf_watermark = PdfFileReader(pdf_buffer)
-            except:
-                logger.exception('Failed to load watermark')
-
-        if not pdf_watermark:
-            logger.error(
-                'No usable watermark found, got %s...', watermark[:100]
-            )
-            return result
-
-        if pdf_watermark.numPages < 1:
-            logger.error('Your watermark pdf does not contain any pages')
-            return result
-        if pdf_watermark.numPages > 1:
-            logger.debug('Your watermark pdf contains more than one page, '
-                         'all but the first one will be ignored')
+            image = Image.open(StringIO(watermark))
+            pdf_buffer = StringIO()
+            if image.mode != 'RGB':
+                image = image.convert('RGB')
+            if report.paperformat_id.format and \
+                    report.paperformat_id.format.lower() in SIZES and \
+                    report.pdf_watermark_scale:
+                image = self._scale_watermark(report, image)
+            image.save(pdf_buffer, 'pdf', quality=100)
+            pdf_watermark = PdfFileReader(pdf_buffer)
+        return pdf_watermark
 
+    @api.model
```

Comment:
read this in the `report` module, and reconsider

Reply:
What is wrong with this? The function seems to be called properly, and the `api.model` is the almost the functional equivalent of `api.v7` and `api.v8`.

Reply:
we don't need an almost functional equivalent, but the correct override, otherwise the function doesn't do the right thing wenn called from v7 or the other way around

Reply:
See this https://github.com/OCA/reporting-engine/pull/283/files#diff-ea48341e81bea8a710c5781b95d7ff8dR83. I call the function both ways and it works correctly. Also, I am not sure with the linter fails and the tests do not run but this module does not seem to be the issue.